### PR TITLE
Kronecker-sum HSGP: non-separable time x graph GP for geo-panel forecasting

### PR DESCRIPTION
**Pull request by juanitorduz**
## Summary

New tutorial notebook `Python/kronecker_hsgp.ipynb` on the Kronecker-sum HSGP: a non-separable Gaussian process on the product space *interval x graph*, built from the HSGP time basis and the eigenbasis of the graph Laplacian. Implemented in NumPyro with `numpyro-forecast`.

Contents:
- Motivation: Gaussian processes for a geo panel, why the exact version does not scale, the HSGP approximation on an interval, and the move to the product space.
- Theory: Matern spectral density and where its power sits, HSGP recap with the functional-calculus and Mercer argument, graph Laplacian and roughness, Kronecker product vs Kronecker sum, effective lengthscales, trace normalization, node standardization, basis size rule.
- Graph of the 16 German states derived from their polygons, Laplacian eigenvectors on the map.
- Prior checks at fixed hyperparameters: spectral weights, non-separability (lag-dependent spatial correlation), node variance before and after standardization, prior draws with identical coefficients.
- Data generating process as a NumPyro model fixed with `handlers.do` (Matern-5/2 trend, graph Ornstein-Uhlenbeck deviations, Fourier seasonality, Student-t noise).
- Three kernels with identical hyperparameters (`independent`, `separable`, `kron_sum`) fitted with NUTS, diagnostics, basis-size check with an `m = 96` refit.
- 13-week forecasts, CRPS / RMSE / coverage, CRPS by horizon, spatial correlation of the forecast draws by horizon.
- Rolling-origin backtest (6 windows) with paired CRPS differences.

Results on the synthetic panel: `kron_sum` has the lowest CRPS in every backtest window. The mean paired difference against `separable` is `-0.130` and against `independent` it is `-0.118`. The `separable` kernel is slightly *worse* than `independent` (mean `+0.012`), so the whole gain is attributable to non-separability rather than to spatial pooling.

## Other changes

- `pixi.toml` / `pixi.lock`: add `numpyro-forecast >=0.3.0,<0.4`.
- `pyproject.toml`: ignore pep8-naming (`N`) for this notebook, since `U` and `R` follow the math (same precedent as `bg_nbd_pymc.ipynb`).
- `AGENTS.md`: writing-style rules for voice, sentences, formatting and structure.

## Notes

Two issues found while porting the original scaffold, both handled in the notebook:
1. The time box must be fixed from the full duration (train + horizon) at model construction, and each call slices the prefix of the basis. Building the box from the length of the current call gives different bases at training and forecast time.
2. The per-region standardization is folded into the eigenvector matrix (`U * scale[:, None]`), which keeps the block as two matrix multiplications and avoids a JAX/XLA CPU miscompile observed with the trailing broadcast multiply under `jit(vmap(...))`.

The notebook runs end to end in about 6 minutes on an M4 Pro:

```bash
cd Python && pixi run jupyter nbconvert --to notebook --execute --inplace kronecker_hsgp.ipynb
```

### CODE REVIEW

**Comment by juanitorduz** [comment]
@claude review

**Comment by juanitorduz** [comment]
Review comments addressed in c311ffa (notebook re-executed end to end, ~5 min, no errors; diagnostics clean except two divergences out of 4000 in `kron_sum`, noted in the text).

| Thread (cell) | Change |
|---|---|
| Intro (0) | Rewrote the eigenvalue sentence (eigenfunctions are products, eigenvalues are sums). HSGP intro post linked in the first paragraph. Added PyMC HSGP-Advanced example 2, NYC BYM, Laplacian eigenmaps post + PyData 2018 slides, GPML ch. 4 (inline and in References). |
| Estimand (5) | One paragraph per score: CRPS with its formula and link to the CRPS post, RMSE, coverage (what 0.9 means). NYC BYM cited where the graph is introduced as spatial distance. |
| Theory (6) | Derivation chain: Bochner + Matérn spectral density (GPML eq. 4.15) -> HSGP expansion -> Laplacian quadratic form (roughness) -> Kronecker sum eigenpairs (two-line proof) -> `W_jv` by substitution, plus the per-mode Matérn interpretation worked out. New plots: first basis functions, `S(sqrt(lambda_j))` for three lengthscales, HSGP vs exact Matérn-3/2 kernel, `W_jv` as curves per graph mode for the three priors. Separable vs Kronecker-product (PyMC example) contrast made explicit. |
| Graph (7) | Adjacency is now derived from the state polygons (`data/germany_states.geojson`, isellsoap/deutschlandGeoJSON, dl-de/by-2-0, cited). Graph drawn on the map; eigenvectors drawn on the map. |
| Laplacian (10) | `jnp` throughout (`jnp.linalg.eigh`); eigenvalues as a stem plot instead of a table. |
| Building blocks (14) | One function per cell with a markdown motivation. `spectral_weights` = `matern_density` + `combine_spectra` (the only place the priors differ) + `trace_normalize`. |
| Fixed hyperparameters (16) | PreliZ plot of the three hyperparameter priors; fixed values stated as being in their bulk. |
| Effective lengthscales (20) | Bar plot of `ell_v` per mode with the `ell` reference line instead of the table. |
| Non-separability (23) | Split into masks/standardization, correlation computation (with the factorization formula that motivates the ratio), and plot. |
| Prior draws (28) | Split into model, draws, plot; explains `substitute` and why the three draws share the coefficients. |
| DGP (31) | Full DGP in LaTeX (trend GP, graph OU with its spectral AR(1) and stationary covariance, seasonality, observations). `panel_dgp` composed from `matern52_covariance` and `graph_ou_covariance`; true values as a table; new plot of the OU decay and stationary sd per graph mode. |
| Summaries (44) | Replaced the 14-row table by `az.plot_dist` per variable across the three models plus the `length`/`rho` forest plot; the `m = 48` vs `m = 96` comparison is also a forest plot. A compact `az.summary` for `kron_sum` remains. |

Note: the PR diff still shows `Python/mrp_pymc.ipynb` because local `master` has the unpushed commit `6d71b99 better dgp` and this branch was cut from it; pushing `master` removes it from the diff.

**Comment by juanitorduz** [comment]
Second review round addressed in 51fdc3a. The notebook was re-executed end to end (6 min, no errors) and every number is unchanged, because no new cell consumes the PRNG stream.

### Comments of 2026-09-18

| Thread (cell) | Change |
|---|---|
| Intro (0) | Rewritten as a ramp: Gaussian processes and what a kernel encodes, why the exact version does not scale, the HSGP as the approximation, what the basis looks like on an interval, and only then the product space. The graph is now introduced as a way to encode proximity ("To encode the proximity of the states we use a network"). The three kernel bullets are expanded to four or five plain sentences each, ending on the consequence for a forecast. The outline is a numbered list in non-technical language whose items match the section headings, starting with "We recall the main concepts of the HSGP approximation on an interval and on a graph". |
| Matern power (4) | New third panel: the fraction of the total power below a given frequency, against `omega / kappa`, for `nu` in {1/2, 3/2, 5/2}. About 82% of the power of a Matern-3/2 process sits below `kappa`, against 50% for 1/2 and 92% for 5/2. The prose explains why: below `kappa` the density is flat because `kappa^2 + omega^2` is essentially `kappa^2`, above it the `omega^2` term takes over and the density decays like `omega^{-(2 nu + 1)}`. Added a remark with four reasons for the Matern family (closed-form spectral density, `nu` controls path smoothness, `nu = 1/2` is the OU process of the DGP, and it is what both HSGP papers and Borovitskiy et al. use). |
| Functional calculus (6) | Added the parenthetical you asked for, on the Dirichlet Laplacian being self-adjoint on `L^2([-L, L])`. Replaced the cryptic one-liner by an explicit two-step argument: the Laplacian eigenfunctions are kernel eigenfunctions with eigenvalue `S(sqrt(lambda_j))`, Mercer writes the kernel as the sum over its eigenpairs, and truncating that sum is the approximation. The paragraph ends by noting that nothing in the argument uses the interval, which is why the same recipe works on a graph. |
| Graph Laplacian (16) | "Rough" is now defined: the quadratic form runs over the edges and adds the squared difference between the endpoints, so it is zero for a constant vector and large when neighbors disagree. The interval analogy uses a distinct symbol, `integral (f'(t))^2 dt`, and says explicitly that the graph form replaces the derivative of `f` by the difference of `x` along an edge. The blue/red sentence now has a plot to refer to: a new figure of the Laplacian eigenvectors of a 10-node path graph, which are discrete cosines and therefore the exact graph analogue of the sine basis plotted a few cells above. The bridge paragraph is rewritten explicitly ("We now have two Laplacians: the Dirichlet Laplacian on the time box and the graph Laplacian...") and the HSGP-Advanced link is removed from here, since it belongs to the Kronecker product section. |
| Kronecker product (17) | Opens by returning to the running example instead of "We now have two eigenbases". States that `P (x) Q` has size `mR x mR`, and explains the indexing: the block in block-row `j` and block-column `j'` is the whole of `Q` scaled by the single number `P_{jj'}`, which is why `Q` carries no index. |
| Scale of the problem (21) | Rewritten as a full sentence with a stated purpose: we move from the 3x3 toy example to the sizes of our own problem so the plots show realistic magnitudes. |
| Pooling strength (25) | The crossover is expanded: `rho mu_v` is what the graph adds next to the 1 contributed by the temporal corner frequency, a mode with `rho mu_v << 1` is treated almost like the constant mode, a mode with `rho mu_v >> 1` is pushed into the tail, and the crossover sits at `mu_v = 1 / rho`. The "pooling strength, not a mixing weight" sentences are removed. |
| Density evaluation (27, 28) | The two adjacent markdown cells are merged into one, opening with "We want to understand the structural difference between the Kronecker product and the Kronecker sum by looking at the weights and the frequencies of the two HSGP approximations". |
| Two things follow (30) | Now a bullet list with bold leads, **Less variance** and **Shorter memory**. |
| Three consequences (33) | Now reads "and we will see all three in a plot below". |
| Shrinkage (35) | Expanded into four paragraphs: a mode with lengthscale `ell_v` carries information for about `ell_v` weeks past the last observation, rough modes decay first, after a few months only the national mode is left, and the visible effect is a state forecast that starts near its own level and moves toward the national one. Points forward to the section that measures it. |
| Heavy sections (41) | Split into four cells: the product-vs-sum recap with the honest remark on the exponent, `### Normalization`, `### Node Standardization`, and `### Basis Size`. Normalization gets a new figure showing the share of the variance carried by each graph mode for four values of `rho` (the constant mode reaches about 90% at `rho = 10`) next to the average variance before and after normalization, which makes the "redistributes rather than shrinks" claim visible. The basis-size figure now has a follow-up sentence. |
| numpy in the graph cell (44) | The adjacency is now built and manipulated in JAX. `geopandas` still returns numpy at the file boundary, and the text says so. Same 29 edges and the same degree vector. |
| Forest plot (125) | The density overlay is replaced by a single `az.plot_forest` across the three models, with the forest column on a log x-axis so that `sigma` (about 0.4) and `length` (about 20 weeks) are readable together. `rho` appears only for the two models that have it. The separate length/rho forest is now redundant and was removed. |
| polars (128) | `az.summary` is displayed directly. |
| print (136) | Removed. The shape is stated in the following sentence. |
| CRPS bar plot (140) | Added, two panels (full horizon and weeks 10 to 13) over the four fitted models, with the values annotated. |
| Legend overlap (143) | Both per-axes legends replaced by one figure-level legend below the two panels. |
| Backtest summary (153) | Added a plot of the paired differences per window, one line per contrast, with a zero line and the mean of each contrast as a dashed line. |

### Threads of 2026-09-14 still marked open

All seven were in fact addressed in c311ffa and are in the notebook today: the adjacency derived from the GeoJSON and drawn on the map (old cell 7), `spectral_weights` split into `mode_hyperparameters`, `unnormalized_weights` and `trace_normalize` (old 14), the PreliZ prior plot (old 16), the `ell_v` bar plot replacing the table (old 20), the split non-separability and prior-draw cells (old 23 and 28), and the forest plots replacing the 14-row summary (old 44). They can be closed.

### Found while going through the notebook

- The claim that "the widest and narrowest bands differ by a factor of about four" was wrong by a square root. The prior *variances* range from 0.53 (Lower Saxony) to 1.99 (Saarland), a ratio of 3.8, so the bands, being standard deviations, differ by about two. Fixed in both places where it appeared.
- RMSE and coverage were reported in two tables but never defined. The expansion asked for in the 2026-09-14 round had been lost. There is now a `### Scoring a Forecast` subsection with one paragraph per score and the CRPS post linked inline.
- Typos: "the observation scal", "to stablish notation", "this contruction", "to our both Laplacians", and a stray `#` in the middle of a paragraph.
- Two cells said the same thing about the box padding, and one cell was an orphan stub pointing at a figure by its title. Merged and removed.
- `### Spectral Weights` and `### Effective Lengthscales` each appeared twice. The ones under "What the Prior Says" are now `### Weights at the Fixed Hyperparameters` and `### Memory of Each Mode`.
- Style pass against the rules this PR adds to `AGENTS.md`: a lead-in sentence for every code cell and a follow-up for every figure and table, no semicolons in prose, structural bold run-in headings turned into real `####` headings (`**Remark.**` kept, it has 111 precedents in the repo), numbers in math mode.
- The conclusion is restructured into `### Key Findings`, `### Model Limitations`, `### Recommendations` and `## Next Steps`.
- The PR description said the mean paired difference against `separable` was -0.028. The notebook output is -0.130. The description is updated.

### One open question

The `AGENTS.md` block added by this PR asks for `{note}` admonitions and `{ref}` cross-references. There are zero occurrences of either in any notebook under `Python/`, and they will not render on the blog. I did not introduce them. Worth either dropping those two rules or noting that they are aspirational.

**Comment by juanitorduz** [comment]
@claude review
